module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1466 @@
+ * second 128 bits of blk_cksum, leaving the first 128 bits for a truncated
+ * regular checksum of the ciphertext which can be used for scrubbing.
+ *
+ * ZIL ENCRYPTION:
+ * ZIL blocks have their bp written to disk ahead of the associated data, so we
+ * cannot store encryption paramaters there as we normally do. For these blocks
+ * the MAC is stored in the embedded checksum within the zil_chain_t header. The
+ * salt and IV are generated for the block on bp allocation instead of at
+ * encryption time. In addition, ZIL blocks have some pieces that must be left
+ * in plaintext for claiming while all of the sensitive user data still needs to
+ * be encrypted. The function zio_crypt_init_uios_zil() handles parsing which
+ * which pieces of the block need to be encrypted.
+ *
+ * DNODE ENCRYPTION:
+ * Similarly to ZIL blocks, the core part of each dnode_phys_t needs to be left
+ * in plaintext for scrubbing and claiming, but the bonus buffers might contain
+ * sensitive user data. The function zio_crypt_init_uios_dnode() handles parsing
+ * which which pieces of the block need to be encrypted.
+ *
+ * CONSIDERATIONS FOR DEDUP:
+ * In order for dedup to work, blocks that we want to dedup with one another
+ * need to use the same IV and encryption key, so that they will have the same
+ * cyphertext. Normally, one should never reuse an IV with the same encryption
+ * key or else AES-GCM and AES-CCM can both actually leak the plaintext of both
+ * blocks. In this case, however, since we are using the same plaindata as
+ * well all that we end up with is a duplicate of the original data we already
+ * had. As a result, an attacker with read access to the raw disk will be able
+ * to tell which blocks are the same but this information is already given away
+ * by dedup anyway. In order to get the same IVs and encryption keys for
+ * equivalent blocks of data we use a HMAC of the plaindata. We use an HMAC
+ * here so there is never a reproducible checksum of the plaindata available
+ * to the attacker. The HMAC key is kept alongside the master key, encrypted
+ * on disk. The first 64 bits of the HMAC are used in place of the random salt,
+ * and the next 96 bits are used as the IV.
+ */
+
+zio_crypt_info_t zio_crypt_table[ZIO_CRYPT_FUNCTIONS] = {
+	{"",			ZC_TYPE_NONE,	0,	"inherit"},
+	{"",			ZC_TYPE_NONE,	0,	"on"},
+	{"",			ZC_TYPE_NONE,	0,	"off"},
+	{SUN_CKM_AES_CCM,	ZC_TYPE_CCM,	16,	"aes-128-ccm"},
+	{SUN_CKM_AES_CCM,	ZC_TYPE_CCM,	24,	"aes-192-ccm"},
+	{SUN_CKM_AES_CCM,	ZC_TYPE_CCM,	32,	"aes-256-ccm"},
+	{SUN_CKM_AES_GCM,	ZC_TYPE_GCM,	16,	"aes-128-gcm"},
+	{SUN_CKM_AES_GCM,	ZC_TYPE_GCM,	24,	"aes-192-gcm"},
+	{SUN_CKM_AES_GCM,	ZC_TYPE_GCM,	32,	"aes-256-gcm"}
+};
+
+static int
+hkdf_sha256_extract(uint8_t *salt, uint_t salt_len, uint8_t *key_material,
+    uint_t km_len, uint8_t *out_buf)
+{
+	int ret;
+	crypto_mechanism_t mech;
+	crypto_key_t key;
+	crypto_data_t input_cd, output_cd;
+
+	/* initialize sha 256 hmac mechanism */
```

Comment:
Yeah. I didn't do this by default because I didn't want to create a dependency on the SHA512 feature flag. Should I make this a dependency or just use SHA512 anyway since its usage is kind of baked into the encryption feature flag.